core/src/main/java/tc/oc/pgm/map/source/GitMapSourceFactory.java

```diff
@@ -0,0 +1,152 @@
+
+  private static File getFile(String source) {
+    credentialsProvided = isCredentialsProvided(source);
+
+    int uniqueIdentifier =
+        Math.abs(source.hashCode()); // Prevents errors on similar repository names
+
+    String dirName;
+    if (credentialsProvided)
+      dirName =
+          (source.substring(source.lastIndexOf("/"), source.indexOf(";")) + "_" + uniqueIdentifier);
+    else dirName = (source.substring(source.lastIndexOf("/") + 1) + "_" + uniqueIdentifier);
+
+    return new File("./gitMaps/" + dirName);
+  }
+
+  public GitMapSourceFactory(String source, Logger logger, boolean updateOnRestart)
+      throws MapMissingException, ClassNotFoundException {
+    super(getFile(source).getPath());
+
+    if (credentialsProvided) this.gitURI = source.substring(0, source.indexOf(";"));
+    else this.gitURI = source;
+
+    this.logger = logger;
+    this.dir = getFile(source);
+    this.updateOnRestart = updateOnRestart;
+
+    // Disable if JGit is not present
+    if (Class.forName("org.eclipse.jgit.api.Git") == null) {
+      throw new MapMissingException(
+          dir.getPath(), "Unable to load JGit(was it excluded when compiling the jar?)");
+    }
+
+    File masterDir = new File("./gitMaps");
+    if (!masterDir.exists()) masterDir.mkdir();
+
+    // Ensure the git map directory exists before trying to update it
+    if (!dir.exists()) {
+      this.updateOnRestart = true;
+      dir.mkdir();
+    }
+    if (!dir.isDirectory())
+      throw new MapMissingException(
+          dir.getPath(), "Unable to read git directory " + dir.getName() + "(Is it a file?)");
+
+    if (credentialsProvided) {
+      provider =
+          new UsernamePasswordCredentialsProvider(
+              source.substring(source.indexOf(";") + 1, source.lastIndexOf(";")),
+              source.substring(source.lastIndexOf(";") + 1));
+    }
+  }
+
+  @Override
+  public Iterator<? extends MapSource> loadNewSources() throws MapMissingException {
+    final Stream<String> paths;
+
+    try {
+      if (updateOnRestart) refreshRepo();
+    } catch (GitAPIException e) {
+      logger.log(
+          Level.WARNING,
+          "Unable to fetch repo " + dir.getName() + ", caused by: " + e.getMessage());
+      if (e instanceof TransportException) {
+        logger.log(Level.INFO, "Provide your username and password in the config like this:");
+        logger.log(Level.INFO, "URL;USERNAME;PASSWORD");
+        logger.log(Level.INFO, "e.g: Github.com/KingOfSquares/maps;KingOfSquares;password1234");
+      }
+    }
+
+    // Load maps separately so it still loads downloaded maps if git connection fails
+    try {
+      paths = loadAllPaths();
+    } catch (IOException e) {
+      throw new MapMissingException(dir.getPath(), "Unable to list files in" + dir.getName(), e);
+    }
+
+    return paths.parallel().flatMap(this::loadNewSource).iterator();
+  }
+
+  private void refreshRepo() throws GitAPIException, MapMissingException {
```

Comment:
This needs to be async.

Reply:
But sync on server startup?

Reply:
Another question, its already called async in `MapLibraryImpl` so why would it not be async already?

Reply:
See the documentation for `MapSourceFactory`. The `Iterator` needs to be "lazy", which means it returns immediately, but will start blocking on the first `next()`